pip/backend.py

```diff
@@ -0,0 +1,293 @@
+import os
+import sys
+import logging
+
+from sysconfig import get_paths
+
+from pip.utils.setuptools_build import SETUPTOOLS_SHIM
+from pip.utils import call_subprocess, ensure_dir
+from pip.utils.temp_dir import TempDirectory
+from pip._vendor import pytoml, six
+
+
+logger = logging.getLogger(__name__)
+
+
+
+class BuildEnvironment(object):
+    """Context manager to install build deps in a simple temporary environment
+    """
+
+    def __init__(self, no_clean):
+        self._temp_dir = TempDirectory(kind="build-env")
+        self._no_clean = no_clean
+        self._temp_dir_created = False
+
+    def __enter__(self):
+        if not self._temp_dir_created:
+            self._temp_dir.create()
+            self._temp_dir_created = True
+
+        self.save_path = os.environ.get('PATH', None)
+        self.save_pythonpath = os.environ.get('PYTHONPATH', None)
+
+        install_scheme = 'nt' if (os.name == 'nt') else 'posix_prefix'
+        install_dirs = get_paths(install_scheme, vars={
+            'base': self._temp_dir.path,
+            'platbase': self._temp_dir.path,
+        })
+
+        scripts = install_dirs['scripts']
+        if self.save_path:
+            os.environ['PATH'] = scripts + os.pathsep + self.save_path
+        else:
+            os.environ['PATH'] = scripts + os.pathsep + os.defpath
+
+        if install_dirs['purelib'] == install_dirs['platlib']:
+            lib_dirs = install_dirs['purelib']
+        else:
+            lib_dirs = install_dirs['purelib'] + os.pathsep + \
+                install_dirs['platlib']
+        if self.save_pythonpath:
+            os.environ['PYTHONPATH'] = lib_dirs + os.pathsep + \
+                self.save_pythonpath
+        else:
+            os.environ['PYTHONPATH'] = lib_dirs
+
+        return self._temp_dir.path
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if not self._no_clean:
+            self._temp_dir.cleanup()
+        if self.save_path is None:
+            os.environ.pop('PATH', None)
+        else:
+            os.environ['PATH'] = self.save_path
+
+        if self.save_pythonpath is None:
+            os.environ.pop('PYTHONPATH', None)
+        else:
+            os.environ['PYTHONPATH'] = self.save_pythonpath
+
+
+class BuildBackend(object):
+    """
+    PEP 517 Build Backend
+    
+    Controls all setup.py interactions
+    """
+
+    def __init__(self, req_install):
+        # The 'pure' OOP way would  be to hold the ONLY reference to
```

Comment:
If there's a comment in first person, I'd suggest that you mention your name or GH username here so that it can be traced to you.

Reply:
Done.